src/devices/ant_antplus.c

```diff
@@ -0,0 +1,198 @@
+
+This decoder only captures and displays the low level packets, identifying the
+2 byte network key plus all other device characteristics and 8 byte ANT payload.
+It identifies ANT+ packets using the unique network key used (0xa6c5)
+Refer to ANT+ documentation for ech specific device to parse the content of
+the pages sent as 8 byte ANT+ payload.
+
+The ANT protocol is using an uncommon strategy for the premable: either 0x55 or 0xaa,
+depending on the value of the first bit of the following byte (0 and 1 respectively)
+nRF24L01, on which the ANT protocol is based, uses the same premable strategy.
+In order to determine if the packet is using 0x55 or 0xaa, both packets are extracted
+and the only valid one is determined by matching the packet CRC with a calculated value
+
+The unpacked payload is 18 bytes long structured as follows:
+
+    PNNDDTXLPPPPPPPPCC
+
+- P: Preamble: either 0x55 or 0xAA, depending on the value of first bit of the next byte
+- N: Network key, assume LSB first (ANT+ uses 0xc5a6, most invalid keys 0x255b)
+- D: Device number, 16 bit. LSB first
+- X: Transmission type
+- L: ANT payload length including CRC
+- P: 8 byte ANT or ANT+ payload
+- C: 16 bit CRC (CRC-16/CCITT-FALSE)
+
+L is always 10, because at the moment the ANT payload is always 8 bytes
+
+CREDITS:
+https://github.com/sghctoma/antfs-poc-defcon24
+https://reveng.sourceforge.io/ to reverse engineer the CRC algorithm used
+*/
+
+#include "decoder.h"
+
+static uint16_t crc16ccitt(uint8_t *data, size_t length)
+{
+    int i;
+    uint16_t crc = 0xffff;
+
+    while (length--) {
+        crc ^= *(unsigned char *)data++ << 8;
+        for (i = 0; i < 8; ++i)
+            crc = crc & 0x8000 ? (crc << 1) ^ 0x1021 : crc << 1;
+    }
+    return crc & 0xffff;
+}
+
+static int ant_antplus_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    uint8_t const preamble[] = {0xAA};
+    uint8_t b1[17], b2[17]; // aligned packet data for both preambles/offsets
+    unsigned bit_offset;
+
+    data_t *data;
+
+    // validate buffer: ANT messages are shorter than 150us, i.e. ~140 bits at 1Mbps
+    if (bitbuffer->bits_per_row[0] < 137 || bitbuffer->bits_per_row[0] > 160) {
+        return DECODE_ABORT_LENGTH;
+    }
+
+    // find a data package and extract data buffer
+    bit_offset = bitbuffer_search(bitbuffer, 0, 0, preamble, sizeof(preamble) * 8) + sizeof(preamble) * 8;
+    if (bit_offset + sizeof(b1) * 8 > bitbuffer->bits_per_row[0]) { // did not find a big enough package
+        return DECODE_ABORT_LENGTH;
+    }
+
+    // ANT and ANT+ packes have either aa or 55 preamble, depending on the first bit of
+    // the following byte. The best way to know which is being used, is to rely on the CRC
+    bitbuffer_extract_bytes(bitbuffer, 0, bit_offset, b1, sizeof(b1) * 8);     // if preamble is AA
+    bitbuffer_extract_bytes(bitbuffer, 0, bit_offset + 1, b2, sizeof(b2) * 8); // if preamble is 55, read starting 1 bit further right
```

Comment:
Check this first. If the crc is valid then continue with the code. If not then extract one more bit and check this crc. If it passes then continue with the code and if not return error. Look in other decoder how to do that.

Reply:
I was struggling on how to structure this. In the beginning I was printing both alternatives to help me understand how it worked, and I kinda got stuck in that mindset. But you are absolutely right, your solution is much more elegant. Also, now that I think of it, most devices that users are likely to find are ANT+, and that uses aa as preamble. So testing the aa alternative first is likely to be successful, avoiding extra code execution.